lib/Infer/AbstractInterpreter.cpp

```diff
@@ -1,6 +1,268 @@
+
+  namespace BinaryTransferFunctionsKB {
+    llvm::KnownBits add(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      return llvm::KnownBits::computeForAddSub(/*Add=*/true, /*NSW=*/false,
+                                               lhs, rhs);
+    }
+
+    llvm::KnownBits addnsw(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      return llvm::KnownBits::computeForAddSub(/*Add=*/true, /*NSW=*/true,
+                                               lhs, rhs);
+    }
+
+    llvm::KnownBits sub(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      return llvm::KnownBits::computeForAddSub(/*Add=*/false, /*NSW=*/false,
+                                               lhs, rhs);
+    }
+
+    llvm::KnownBits subnsw(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      return llvm::KnownBits::computeForAddSub(/*Add=*/false, /*NSW=*/true,
+                                               lhs, rhs);
+    }
+
+    llvm::KnownBits mul(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(lhs.getBitWidth());
+
+      auto trailingZeros0 = lhs.countMinTrailingZeros();
+      auto trailingZeros1 = rhs.countMinTrailingZeros();
+      Result.Zero.setLowBits(std::min(trailingZeros0 + trailingZeros1, lhs.getBitWidth()));
+
+      // check for leading zeros
+      auto lz0 = lhs.countMinLeadingZeros();
+      auto lz1 = rhs.countMinLeadingZeros();
+      auto confirmedLeadingZeros = lz0 + lz1 - 1;
+      auto resultSize = lhs.getBitWidth() + rhs.getBitWidth() - 1;
+      if (resultSize - confirmedLeadingZeros < lhs.getBitWidth())
+	Result.Zero.setHighBits(lhs.getBitWidth() - (resultSize - confirmedLeadingZeros));
+
+      return Result;
+    }
+
+    llvm::KnownBits udiv(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(lhs.getBitWidth());
+      const auto width = Result.getBitWidth();
+
+      unsigned LeadZ = lhs.countMinLeadingZeros();
+      unsigned RHSMaxLeadingZeros = rhs.countMaxLeadingZeros();
+      if (RHSMaxLeadingZeros != width)
+        LeadZ = std::min(width, LeadZ + width - RHSMaxLeadingZeros - 1);
+      Result.Zero.setHighBits(LeadZ);
+      return Result;
+    }
+
+    llvm::KnownBits urem(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(lhs.getBitWidth());
+      const auto width = Result.getBitWidth();
+
+      if (rhs.isConstant()) {
+        auto RA = rhs.getConstant();
+        if (RA.isPowerOf2()) {
+          auto LowBits = (RA - 1);
+          Result = lhs;
+          Result.Zero |= ~LowBits;
+          Result.One &= LowBits;
+          return Result;
+        }
+      }
+
+      unsigned Leaders =
+        std::max(lhs.countMinLeadingZeros(), rhs.countMinLeadingZeros());
+      Result.resetAll();
+      Result.Zero.setHighBits(Leaders);
+      return Result;
+    }
+
+    llvm::KnownBits and_(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      auto result = lhs;
+
```

Comment:
don't put a blank line here if you don't have an analogous blank line in the next function